afflux.1
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "AFFLUX" "1" "April 23, 2023" "Afflux Manual" ""
.hy
.SH NAME
.PP
afflux - AFF4 logical imager
.SH SYNOPSIS
.PP
afflux [-hvrx] [-t DIRECTORY] [-A] [-e PASSWORD] [-o OUTPUT_FILE] [-m
AFF4_IMAGE] [-p PLUGIN_NAME] [-x AFF4_IMAGE] [PLUGIN_OPTIONS]
.SH DESCRIPTION
.PP
Afflux is an AFF4 logical imager.
The goal of the tool is to logically image modern devices when a
physical image may not be possible.
Afflux supports several optional arguments no matter which plugin is
selected.
These are shown with \[lq]afflux -h\[rq] and are also shown with each
plugin\[cq]s \[lq]-h\[rq] argument.
.SH OPTIONS
.TP
\f[B]-h\f[R], \f[B]\[en]help\f[R]
Print help output.
.TP
\f[B]-v\f[R], \f[B]\[en]verbose\f[R]
Enable verbose output.
.TP
\f[B]-t PATH\f[R], \f[B]\[en]temp PATH\f[R]
Directory to create any temporary files.
.TP
\f[B]-A\f[R], \f[B]\[en]append\f[R]
Append to an existing AFF4 image specified with \[lq]-o\[rq].
.TP
\f[B]-e PASSWORD\f[R], \f[B]\[en]container_password PASSWORD\f[R]
Password to create an encrypted AFF4 container.
.TP
\f[B]-o OUTPUT_FILE\f[R], \f[B]\[en]output OUTPUT_FILE\f[R]
Output AFF4 file name.
\f[B]\[en]overwrite\f[R]
Overwrite AFF4 file if it already exists.
.TP
\f[B]-m AFF4_IMAGE\f[R], \f[B]\[en]meta AFF4_IMAGE\f[R]
Print the AFF4 image\[cq]s metadata.
.TP
\f[B]-p PLUGIN_NAME\f[R], \f[B]\[en]plugin PLUGIN_NAME\f[R]
Specify a plugin to load, or use \[lq]list\[rq] to list all plugins.
.TP
\f[B]-z\f[R], \f[B]\[en]zip\f[R]
Write to a Zip container instead of AFF4.
.TP
\f[B]-x AFF4_IMAGE\f[R], \f[B]\[en]extract AFF4_IMAGE\f[R]
Extract all files and folders from an AFF4 image.
.SH DEFAULT PLUGINS
.TP
\f[B]android_adb\f[R]
Image an Android device via ADB over USB or over network.
Supports root filesystems.
.TP
\f[B]apple_afc\f[R]
Image a device that support AFC.
Supports jailbroken and non-jailbroken devices.
.TP
\f[B]disk_image\f[R]
Image a file or folder on the local disk.
.TP
\f[B]generic_ftp\f[R]
Image a device via network FTP
.TP
\f[B]generic_http\f[R]
Image a file or folder via http.
.TP
\f[B]generic_smb\f[R]
Image a device via network SMB.
.TP
\f[B]generic_ssh\f[R]
Image a device via network SSH.
.TP
\f[B]ios_ssh\f[R]
Image a jailbroken device via SSH over network or USB.
.TP
\f[B]usb_drive\f[R]
Image a local USB device or mounted drive.
.SH Apple_AFC Plugin
.PP
The \[lq]apple_afc\[rq] plugin allows for creating logical images from
Apple devices using the AFC protocol.
The iPod Touch, iPhone, and Apple Watch are supported.
\[lq]-i\[rq] is used to specify \[lq]AFC\[rq] for non-jailbroken devices
or \[lq]AFC2\[rq] for jailbroken devices with \[lq]Apple File Conduit
2\[rq] installed.
\[lq]-c\[rq] is used to remove all previous pairing records and
\[lq]-k\[rq] is used to initiate pairing before imaging.
.TP
\f[B]-i METHOD\f[R], \f[B]\[en]iOS METHOD\f[R]
Create an image from an iOS device using METHOD.
Jailbroken methods: AFC, AFC2.
Unjailbroken method: AFC.
.TP
\f[B]-c\f[R], \f[B]\[en]clear\f[R]
Clear any pairing records made by Afflux.
.TP
\f[B]-k\f[R], \f[B]\[en]pair\f[R]
Pair or re-pair the device before imaging.
.SH iOS_SSH Plugin
.PP
The \[lq]ios_ssh\[rq] plugin is used to image a Jailbroken device via
SSH protocol.
\[lq]-a\[rq] is used to specify the devices IP address to connect to or
\[lq]-l\[rq] is used for a device connected via USB.
\[lq]-R\[rq] is used to specify the root password if it is known,
default is \[lq]alpine\[rq].
\[lq]-s\[rq] is used to enable traversing symlinks.
.TP
\f[B]-a IP\f[R], \f[B]\[en]address IP\f[R]
IP address to connect to.
.TP
\f[B]-d REMOTE DIR\f[R], \f[B]\[en]directory REMOTE DIR\f[R]
Remote directory to begin imaging.
.TP
\f[B]-l\f[R], \f[B]\[en]local\f[R]
SSH to iOS device via network address instead of USB tunneling.
.TP
\f[B]-R ROOT_PASSWORD\f[R], \f[B]\[en]root-password ROOT_PASSWORD\f[R]
Root password for jailbroken device if using SSH.
Defaults to \[lq]alpine\[rq].
.TP
\f[B]-s\f[R], \f[B]\[en]symlinks\f[R]
Follow and image any symlinks.
.SH Android_ADB Plugin
.PP
The \[lq]android_adb\[rq] plugin allows for imaging an Android device
over the ADB protocol.
\[lq]-a\[rq] is used to specify the \[lq]IP:PORT\[rq] of the device if
it is on the network or \[lq]USB\[rq] if it is connected via USB.
\[lq]-d\[rq] is used to specify the directory to image.
Default directory is \[lq]/\[rq].
Directories that should be ignored can be specified here by placing a
\[lq]-\[rq] after the directory to ignore.
For example, to image \[lq]/\[rq] and ignore \[lq]/proc\[rq] you would
use \[lq]-d / /proc-\[rq].
\[lq]-s\[rq] is used to enable the imaging to traverse symlinks.
And \[lq]-k\[rq] is used to pair or repair the device.
.TP
\f[B]-a USB/IP:PORT\f[R], \f[B]\[en]android USB/IP:PORT\f[R]
Create an image from a connected Android device using ADB.
Method can be USB or via IP.
.TP
\f[B]-d [DIRECTORY [DIRECTORY \&...]]\f[R], \f[B]\[en]directory [DIRECTORY [DIRECTORY \&...]]\f[R]
Directory to image on the device.
.TP
\f[B]-s\f[R], \f[B]\[en]symlinks\f[R]
Follow and image any symlinks.
.TP
\f[B]\[en]root\f[R]
Check and acquire root privileges if possible.
.TP
\f[B]-k\f[R], \f[B]\[en]keygen\f[R]
Regenerate pairing keys for the device.
.TP
\f[B]\[en]timeout\f[R]
ADB timeout amount.
.TP
\f[B]\[en]auth-timeout\f[R]
Timeout for CNXN authentication response.
.SH Generic_SSH Plugin
.PP
The \[lq]generic_ssh\[rq] plugin is used for imaging any device that
supports a network SSH connection.
\[lq]-a\[rq] is used to specify the IP address.
\[lq]-u\[rq] is used to specify the username.
\[lq]-P\[rq] is the user\[cq]s password.
\[lq]-q\[rq] can specify the port.
The default is \[lq]22\[rq].
\[lq]-s\[rq] will enable symlink traversal and \[lq]-d\[rq] is used to
specify a directory to image.
.TP
\f[B]-a IP\f[R], \f[B]\[en]address IP\f[R]
IP address to connect to.
.TP
\f[B]-u USERNAME\f[R], \f[B]\[en]username USERNAME\f[R]
SSH username to connect with.
.TP
\f[B]-P PASSWORD\f[R], \f[B]\[en]password PASSWORD\f[R]
Password for the SSH user.
.TP
\f[B]\[en]port PORT\f[R]
Port to connect over SSH.
.TP
\f[B]-s\f[R], \f[B]\[en]symlinks\f[R]
Follow and image any symlinks.
.TP
\f[B]-r\f[R], \f[B]\[en]recursive\f[R]
Add files and folders recursively.
.TP
\f[B]-d REMOTE DIR\f[R], \f[B]\[en]directory REMOTE DIR\f[R]
Remote directory to begin imaging.
.SH Generic_FTP Plugin
.PP
The \[lq]generic_ftp\[rq] plugin is used for imaging any device that
supports a network FTP connection.
\[lq]-a\[rq] is used to specify the hostname or IP address.
\[lq]-u\[rq] is used to specify the username.
\[lq]-P\[rq] is the user\[cq]s password.
\[lq]-q\[rq] can specify the port.
The default is \[lq]21\[rq].
\[lq]-s\[rq] will enable symlink traversal and \[lq]-d\[rq] is used to
specify a directory to image.
.TP
\f[B]-a IP\f[R], \f[B]\[en]address IP\f[R]
IP address or server name to connect to.
.TP
\f[B]-u USERNAME\f[R], \f[B]\[en]username USERNAME\f[R]
FTP username to connect with.
.TP
\f[B]-P PASSWORD\f[R], \f[B]\[en]password PASSWORD\f[R]
Password for the FTP user.
.TP
\f[B]\[en]port PORT\f[R]
Port to connect over FTP.
.TP
\f[B]-s\f[R], \f[B]\[en]symlinks\f[R]
Follow and image any symlinks.
.TP
\f[B]-r\f[R], \f[B]\[en]recursive\f[R]
Add files and folders recursively.
.TP
\f[B]-d REMOTE_DIR\f[R], \f[B]\[en]directory REMOTE_DIR\f[R]
Remote directory to begin imaging.
.SH USB_Drive Plugin
.PP
The \[lq]usb_drive\[rq] image is used for imaging a USB drive.
Simply add the flash drive name with \[lq]-u\[rq].
.TP
\f[B]-u USB_NAME\f[R], \f[B]\[en]usb USB_NAME\f[R]
Create image from a connected USB device.
.TP
\f[B]-s\f[R], \f[B]\[en]symlinks\f[R]
Follow and image any symlinks.
.TP
\f[B]\[en]timeout\f[R]
FTP connection timeout.
.SH Disk_Image Plugin
.PP
The \[lq]disk_image\[rq] plugin is used to image files or folders on a
disk.
\[lq]-f\[rq] is used to specify a folder.
\[lq]-F\[rq] is used to specify a file.
And again, \[lq]-s\[rq] is used enable symlink traversal.
.TP
\f[B]-f [FOLDER [FOLDER \&...]]\f[R], \f[B]\[en]folder [FOLDER [FOLDER \&...]]\f[R]
Create image from a local folder.
.TP
\f[B]-F [FILE [FILE \&...]]\f[R], \f[B]\[en]file [FILE [FILE \&...]]\f[R]
Create image from a local file or files.
.TP
\f[B]-s\f[R], \f[B]\[en]symlinks\f[R]
Follow and image any symlinks.
.TP
\f[B]-r\f[R], \f[B]\[en]recursive\f[R]
Add files and folders recursively.
.SH Generic_SMB Plugin
.PP
The \[lq]generic_smb\[rq] plugin is used for imaging SMB shares across
systems.
\[lq]-a\[rq] is used to specify the hostname/address for the share.
\[lq]-u\[rq] is the username and \[lq]-P\[rq] is the password.
The port can also be specified to a non-standard SMB port with
\[lq]-p\[rq].
Symlinks can also be traversed with \[lq]-s\[rq].
\[lq]-S\[rq] is used to specify the share directory to image.
.TP
\f[B]-a HOSTNAME\f[R], \f[B]\[en]address HOSTNAME\f[R]
IP address or server name to connect to.
.TP
\f[B]-S SHARE\f[R], \f[B]\[en]share SHARE\f[R]
SMB share for imaging.
.TP
\f[B]-u USERNAME\f[R], \f[B]\[en]username USERNAME\f[R]
SMB username to connect with.
.TP
\f[B]-P PASSWORD\f[R], \f[B]\[en]password PASSWORD\f[R]
Password for the SMB user.
.TP
\f[B]\[en]port PORT\f[R]
Port to connect over SMB.
.TP
\f[B]-s\f[R], \f[B]\[en]symlinks\f[R]
Enable traversing symlinks.
.TP
\f[B]-r\f[R], \f[B]\[en]recursive\f[R]
Add files and folders recursively.
.SH Generic_HTTP Plugin
.PP
The \[lq]generic_http\[rq] plugin is used for imaging HTTP directories.
\[lq]-l\[rq] is used to specify the link or link(s) to image from.
\[lq]-c\[rq] is used to specify the chunk size for the downloads.
\[lq]-r\[rq] applies to this module if you want to image all the HTTP
directories.
.TP
\f[B]-l [LINK [LINK \&...]]\f[R], \f[B]\[en]link [LINK [LINK \&...]]\f[R]
Create image from a link.
.TP
\f[B]-c [CHUNK_SIZE [CHUNK_SIZE \&...]]\f[R], \f[B]\[en]chunk-size [CHUNK_SIZE [CHUNK_SIZE \&...]]\f[R]
Chunk size to download with.
Default is 1024.
.TP
\f[B]-r\f[R], \f[B]\[en]recursive\f[R]
Add files and folders recursively.
.SH EXAMPLES
.TP
\f[B]afflux -p disk_image -f /home -o home.aff4\f[R]
Create an image (\[lq]home.aff4\[rq]) of the \[lq]/home\[rq] directory
on a Linux system.
.TP
\f[B]afflux -p android_adb -a 10.11.1.5:5001 \[en]root -d / /dev- /proc- -o test.aff4 -v\f[R]
Verbosely image an Android phone (\[lq]10.11.1.5\[rq]) over the network
to \[lq]test.aff4\[rq].
Attempt to get root access, start imaging at \[lq]/\[rq] and ignore
\[lq]/dev\[rq] and \[lq]/proc\[rq].
.TP
\f[B]afflux -p ios_ssh -a 10.11.1.7 -o iphone_ssh.aff4\f[R]
Image a jailbroken iPhone (\[lq]10.11.1.7\[rq]) via SSH.
.TP
**afflux -p generic_smb -a WINDEV2210EVAL -S _folder -o test.aff4 -v -u User -P test**
Image an SMB share on a Windows machine.
.SH COPYRIGHT
.PP
Copyright 2020 Andrew Clark.
License GPLv3+: GNU GPL version 3 or later
<https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH AUTHORS
Andrew Clark IV.


afflux.py
# By Marshall University on 7/10/2021

import argparse
from datetime import datetime
from os import path, chdir, getcwd, environ
from sys import exit

from colorama import init, deinit, Fore

import imaging
from plugin_manager import PluginCollection

init(autoreset=True)

text_art = r"""
_____/\\\\\\\\\_____/\\\\\\\\\\\\\\\__/\\\\\\\\\\\\\\\__/\\\______________/\\\________/\\\__/\\\_______/\\\_        
 ___/\\\\\\\\\\\\\__\/\\\///////////__\/\\\///////////__\/\\\_____________\/\\\_______\/\\\_\///\\\___/\\\/__       
  __/\\\/////////\\\_\/\\\_____________\/\\\_____________\/\\\_____________\/\\\_______\/\\\___\///\\\\\\/____      
   _\/\\\_______\/\\\_\/\\\\\\\\\\\_____\/\\\\\\\\\\\_____\/\\\_____________\/\\\_______\/\\\_____\//\\\\______     
    _\/\\\\\\\\\\\\\\\_\/\\\///////______\/\\\///////______\/\\\_____________\/\\\_______\/\\\______\/\\\\______    
     _\/\\\/////////\\\_\/\\\_____________\/\\\_____________\/\\\_____________\/\\\_______\/\\\______/\\\\\\_____   
      _\/\\\_______\/\\\_\/\\\_____________\/\\\_____________\/\\\_____________\//\\\______/\\\_____/\\\////\\\___  
       _\/\\\_______\/\\\_\/\\\_____________\/\\\_____________\/\\\\\\\\\\\\\\\__\///\\\\\\\\\/____/\\\/___\///\\\_ 
        _\///________\///__\///______________\///______________\///////////////_____\/////////_____\///_______\///__
        
        Version 1.1"""


def main():
    parser = argparse.ArgumentParser(description=f"{Fore.GREEN}AFF4 logical imager.{Fore.RESET}",
                                     formatter_class=argparse.RawTextHelpFormatter, add_help=False)

    # General command line options
    parser.add_argument('-h', "--help", action="store_true",
                        help='print help output.')
    parser.add_argument('-v', "--verbose", action="store_true", default=False,
                        help='enable verbose output.')
    parser.add_argument('-t', "--temp", nargs=1, action="store", metavar='PATH',
                        help='directory to create any temporary files.')
    parser.add_argument('-A', "--append", action="store_true", default=False,
                        help='append to an existing AFF4 image specified with `-o`.')
    parser.add_argument('-e', "--container_password", nargs=1, action="store", metavar='PASSWORD',
                        help='password to create an encrypted AFF4 container.')
    parser.add_argument('-o', "--output", nargs=1, action="store", metavar='OUTPUT_FILE',
                        help='output AFF4 file name.')
    parser.add_argument("--overwrite", action="store_true", default=False,
                        help='overwrite AFF4 file if it already exists.')
    parser.add_argument('-m', "--meta", action="store", nargs=1, metavar='AFF4_IMAGE',
                        help='print the AFF4 image\'s metadata.', )
    parser.add_argument('-p', "--plugin", action="store", nargs=1, metavar='PLUGIN_NAME',
                        help='specify a plugin to load, or use "list" to list all plugins.', )
    parser.add_argument("-z", "--zip", action="store_true", default=False,
                        help='write to a Zip container instead of AFF4.')

    # Temporarily set 'QT_STYLE_OVERRIDE' environment variable to suppress QT warnings on some systems.
    environ["QT_STYLE_OVERRIDE"] = ""

    print(f"{Fore.GREEN}{text_art}\n")
    print(f"{Fore.GREEN}\n\n[+] Afflux Logical Imager Starting...")
    args, extra_args = parser.parse_known_args()
    plugins = PluginCollection('plugins', args.verbose)
    imager = imaging.Imager(args.verbose)
    imager.check_os()
    if args.verbose:
        print(f"{Fore.GREEN}[*] OS: {Fore.RESET}{imager.os}")
    current_path = getcwd()

    if args.plugin is not None:
        parser.add_argument('-x', "--extract", action="store", nargs=1, metavar='AFF4_IMAGE',
                            help=f'extract all files and folders from an AFF4 image.\n\n{Fore.GREEN}[Plugin Options]'
                                 f'{Fore.RESET}\n')
    else:
        parser.add_argument('-x', "--extract", action="store", nargs=1, metavar='AFF4_IMAGE',
                            help='extract all files and folders from an AFF4 image.\n\n')

    args, extra_args = parser.parse_known_args()
    # Print help for '-h' and no plugin argument.
    if args.help and args.plugin is None:
        parser.print_help()
        exit()

    # Also print help if nothing is given.
    if not any(vars(args).values()):
        print(parser.format_help())
        exit()

    # Print container metadata
    if args.meta:
        if path.exists(args.meta[0]):
            print(f"{Fore.GREEN}[+] Loading image...")
            imager.meta(args.meta[0], args.container_password)
            exit()
        else:
            print(f"{Fore.RED}[-] File not found: {Fore.RESET}{args.meta[0]}\n")
            exit()

    # Print help if no -p switch is used.
    if args.plugin is None and args.extract is None:
        parser.print_help()
        exit()

    # If '-p' is 'list', list plugins.
    if args.plugin:
        if args.plugin[0].lower() == 'list':
            plugin_list = plugins.list_plugins()
            print(f"{Fore.GREEN}\n[+] Plugins:\n")
            for plugin in plugin_list:
                print(f"{plugin.name} \t| {plugin.description}")
            print()
            exit()

    # Check to make sure '.aff4' or '.zip' is prepended to the output. If not, add it.
    if args.output:
        if not args.extract:
            if not args.zip:
                if args.output[0][-5:] != ".aff4":
                    args.output = [args.output[0] + ".aff4"]
            else:
                if args.output[0][-4:] != ".zip":
                    args.output = [args.output[0] + ".zip"]

    # TODO: Should we check this here?
    # Check if an output file is given with a plugin.
    if args.plugin:
        if not args.help:
            if not args.output:
                print(f"{Fore.RED}[-] No output file given. Use '-o' to specify an output file. "
                      f"To see options, use '-h'.\n")
                exit()

    # Check if the file already exists
    if args.output:
        if not args.extract:
            if not args.append:
                if not args.overwrite:
                    if path.exists(args.output[0]):
                        print(f"{Fore.GREEN}[*] '{args.output[0]}' already exists. Overwrite? (y/n)")
                        overwrite = input("> ").lower().strip()
                        if (overwrite == "y") or (overwrite == "yes"):
                            pass
                        else:
                            exit()
            else:
                # Check if the file we are appending to exists.
                if not path.exists(args.output[0]):
                    print(f"{Fore.RED}[-] '{args.output[0]}' does not exist.")
                    exit()

    start_time = datetime.now()
    try:
        # Cleanup anything if needed.
        clean = imager.cleanup(specified_path=args.temp)
        if not clean:
            exit()

        # Try and load the specified plugin.
        if args.plugin is not None:
            plugin = plugins.return_plugin(args.plugin[0])
            if plugin is not False:
                plugin.setup_arg_parser(extra_args, args, parser)
                if args.help:
                    plugin.parser.print_help()
                else:
                    plugin.run()
            else:
                print(f"{Fore.RED}[-] Plugin not found.\n")
                exit()

        # Extract files from the container
        elif args.extract:
            print(f"{Fore.GREEN}[*] Preparing to extract.")
            if not path.isfile(args.extract[0]):
                print(f"{Fore.RED}[-] AFF4 file not found: {Fore.RESET}{args.extract[0]}\n")
                exit()
            if not args.output:
                print(f"{Fore.RED}[-] Specify output directory with '-o'.\n")
                exit()
            imager.extract_all(args.extract[0], args.output[0], args.container_password)
        print(f"{Fore.GREEN}\n[*] Time: {Fore.RESET}{str(datetime.now() - start_time)}\n")

    except KeyboardInterrupt:
        print(f"{Fore.GREEN}\n[*] Attempting to quit gracefully...")
        chdir(current_path)
        imager.cleanup(specified_path=args.temp)
        exit()

    except FileExistsError as e:
        if args.verbose:
            print(e)
        print(f"{Fore.RED}\n[-] Need to clean up errors from last run. Please run again...")
        chdir(current_path)
        imager.cleanup(specified_path=args.temp)
        exit()


if __name__ == "__main__":
    main()
    deinit()                # De-initialize colorama
